### ISSUE TITLE
Validation loss is nan

### ISSUE THREAD

**Issue by allipilli-harshitha** [AntonMu/TrainYourOwnYOLO#125]
@AntonMu I am getting nan values for val_loss
I have changed anchors according to my data set using kmeans.py in cfg file
classes=4, filters=27, width=height=608
in yolo_train.py, i have changed batchsize=1, earlystopping(patience=20), input_shape=608,608,
model.compile(optimizer=Adam(lr=1e-10)

in cfg file i have changed last yolo layers mask values from 0,1,2 to 1,2,3.

before changing those parameters also i got val_loss as nan and even after changing alsi i am getting val_loss as nan
pls help me out with this issue
![80602383-ae060880-8a4c-11ea-92db-f748d49c8285](https://user-images.githubusercontent.com/60619885/81398132-42a3f100-9146-11ea-990c-90b58760f1c7.jpeg)

**Comment by AntonMu** [comment]
Hi @allipilli-harshitha - it looks like you changed a lot of parameters. I would try it again by following all the provided instructions using your own dataset. Then once this work you can slowly start changing the model.

**Comment by allipilli-harshitha** [comment]
Hi @AntonMu  for the first time when i trained the model i haven't changed much of the parameters. I have just changed filters and classes, then also i got nan value.
That is why i have changed few more parameters, but still i am getting same nan values for val loss

**Comment by AntonMu** [comment]
Hi @allipilli-harshitha - unless you really understand the implications I wouldn't change the filters. Even for classes you don't need to modify the code. All you need to do is tag several classes in VoTT.

**Comment by allipilli-harshitha** [comment]
@AntonMu  My dataset is car tyres wearout detection. So i can annotate to only 4 classes. I could not do with more.
So if i keep the filters as 255 and classes as 80 also there will be no problem ?

**Comment by AntonMu** [comment]
Yes, simply follow the provided instructions in the Readme and don't change any other code.

**Comment by allipilli-harshitha** [comment]
Classes alsi should i leave it as it is i.e., 80 or can i change it to 4

**Comment by AntonMu** [comment]
I don't know where you get the 80 classes from. In the provided example there is only one class (cat face). But if you want to have several classes, you just need to label more classes.

**Comment by allipilli-harshitha** [comment]
In the cfg file provided by you it is 80 classes. That is what im asking for

**Comment by AntonMu** [comment]
I highly recommend that you follow the instructions in the Readme. That will guarantee success. There is no mention of changing the cfg file and doing so will make you stray off and lead to bugs.

**Comment by allipilli-harshitha** [comment]
Hi @AntonMu  i have not changed any of the code and tried and i did not get nan values.
 i even tried changing the classes to 4 in cfg file, and i did not get nan values. 
Thank you 
I will be closing this issue

**Comment by AntonMu** [comment]
Good to hear!

**Comment by rpentela** [comment]
> In the cfg file provided by you it is 80 classes. That is what im asking for



> Hi @AntonMu i have not changed any of the code and tried and i did not get nan values.
> i even tried changing the classes to 4 in cfg file, and i did not get nan values.
> Thank you
> I will be closing this issue

Hi how do u solve it. Initially you said with classes 4 you got nan

**Comment by allipilli-harshitha** [comment]
> > In the cfg file provided by you it is 80 classes. That is what im asking for
> 
> > Hi @AntonMu i have not changed any of the code and tried and i did not get nan values.
> > i even tried changing the classes to 4 in cfg file, and i did not get nan values.
> > Thank you
> > I will be closing this issue
> 
> Hi how do u solve it. Initially you said with classes 4 you got nan

Hi @rpentela  for first i have changed filters=27 and classes=4. For that i got nan values. 
So without changing anything except classes to 4, i trained then i did not get nan values.

**Comment by rpentela** [comment]
> > > In the cfg file provided by you it is 80 classes. That is what im asking for
> > 
> > > Hi @AntonMu i have not changed any of the code and tried and i did not get nan values.
> > > i even tried changing the classes to 4 in cfg file, and i did not get nan values.
> > > Thank you
> > > I will be closing this issue
> > 
> > Hi how do u solve it. Initially you said with classes 4 you got nan
> 
> Hi @rpentela  for first i have changed filters=27 and classes=4. For that i got nan values. 
> So without changing anything except classes to 4, i trained then i did not get nan values.

How much time it took. Will it work only on given images in this repository or any images

**Comment by allipilli-harshitha** [comment]
> > > > In the cfg file provided by you it is 80 classes. That is what im asking for
> > > 
> > > 
> > > > Hi @AntonMu i have not changed any of the code and tried and i did not get nan values.
> > > > i even tried changing the classes to 4 in cfg file, and i did not get nan values.
> > > > Thank you
> > > > I will be closing this issue
> > > 
> > > 
> > > Hi how do u solve it. Initially you said with classes 4 you got nan
> > 
> > 
> > Hi @rpentela  for first i have changed filters=27 and classes=4. For that i got nan values.
> > So without changing anything except classes to 4, i trained then i did not get nan values.
> 
> How much time it took. Will it work only on given images in this repository or any images

It takes long time to train the model mostly whole day. 
It works on any images.
My dataset contains car tyres and it worked on it.

**Comment by rpentela** [comment]
> > > > > In the cfg file provided by you it is 80 classes. That is what im asking for
> > > > 
> > > > 
> > > > > Hi @AntonMu i have not changed any of the code and tried and i did not get nan values.
> > > > > i even tried changing the classes to 4 in cfg file, and i did not get nan values.
> > > > > Thank you
> > > > > I will be closing this issue
> > > > 
> > > > 
> > > > Hi how do u solve it. Initially you said with classes 4 you got nan
> > > 
> > > 
> > > Hi @rpentela  for first i have changed filters=27 and classes=4. For that i got nan values.
> > > So without changing anything except classes to 4, i trained then i did not get nan values.
> > 
> > How much time it took. Will it work only on given images in this repository or any images
> 
> It takes long time to train the model mostly whole day. 
> It works on any images.
> My dataset contains car tyres and it worked on it.

Thanks. I am working on 6 classes surface defect detection. Started the model today. Will see

**Comment by rpentela** [comment]
> > > > > > In the cfg file provided by you it is 80 classes. That is what im asking for
> > > > > 
> > > > > 
> > > > > > Hi @AntonMu i have not changed any of the code and tried and i did not get nan values.
> > > > > > i even tried changing the classes to 4 in cfg file, and i did not get nan values.
> > > > > > Thank you
> > > > > > I will be closing this issue
> > > > > 
> > > > > 
> > > > > Hi how do u solve it. Initially you said with classes 4 you got nan
> > > > 
> > > > 
> > > > Hi @rpentela  for first i have changed filters=27 and classes=4. For that i got nan values.
> > > > So without changing anything except classes to 4, i trained then i did not get nan values.
> > > 
> > > How much time it took. Will it work only on given images in this repository or any images
> > 
> > It takes long time to train the model mostly whole day. 
> > It works on any images.
> > My dataset contains car tyres and it worked on it.
> 
> Thanks. I am working on 6 classes surface defect detection. Started the model today. Will see

Mine is 4 gb RAM win 7 64 bit. Taking long time. Is it enough

**Comment by allipilli-harshitha** [comment]
> > > > > > > In the cfg file provided by you it is 80 classes. That is what im asking for
> > > > > > 
> > > > > > 
> > > > > > > Hi @AntonMu i have not changed any of the code and tried and i did not get nan values.
> > > > > > > i even tried changing the classes to 4 in cfg file, and i did not get nan values.
> > > > > > > Thank you
> > > > > > > I will be closing this issue
> > > > > > 
> > > > > > 
> > > > > > Hi how do u solve it. Initially you said with classes 4 you got nan
> > > > > 
> > > > > 
> > > > > Hi @rpentela  for first i have changed filters=27 and classes=4. For that i got nan values.
> > > > > So without changing anything except classes to 4, i trained then i did not get nan values.
> > > > 
> > > > 
> > > > How much time it took. Will it work only on given images in this repository or any images
> > > 
> > > 
> > > It takes long time to train the model mostly whole day.
> > > It works on any images.
> > > My dataset contains car tyres and it worked on it.
> > 
> > 
> > Thanks. I am working on 6 classes surface defect detection. Started the model today. Will see
> 
> Mine is 4 gb RAM win 7 64 bit. Taking long time. Is it enough

Mine is 8 gb ram 64-bit os and gpu is very less. It took long time to train like for 1 epoch it took 5 hrs. Imagine 51 to 102 epochs.
So i asked my frnd to train tge model on her lap. Even her lap is having 8 gb ram but she have 4 gb of gpu. So it took 1 day to train.

Better you use somebody else's lap with high gpu or try using cloud platforms

**Comment by rpentela** [comment]
> > > > > > > > In the cfg file provided by you it is 80 classes. That is what im asking for
> > > > > > > 
> > > > > > > 
> > > > > > > > Hi @AntonMu i have not changed any of the code and tried and i did not get nan values.
> > > > > > > > i even tried changing the classes to 4 in cfg file, and i did not get nan values.
> > > > > > > > Thank you
> > > > > > > > I will be closing this issue
> > > > > > > 
> > > > > > > 
> > > > > > > Hi how do u solve it. Initially you said with classes 4 you got nan
> > > > > > 
> > > > > > 
> > > > > > Hi @rpentela  for first i have changed filters=27 and classes=4. For that i got nan values.
> > > > > > So without changing anything except classes to 4, i trained then i did not get nan values.
> > > > > 
> > > > > 
> > > > > How much time it took. Will it work only on given images in this repository or any images
> > > > 
> > > > 
> > > > It takes long time to train the model mostly whole day.
> > > > It works on any images.
> > > > My dataset contains car tyres and it worked on it.
> > > 
> > > 
> > > Thanks. I am working on 6 classes surface defect detection. Started the model today. Will see
> > 
> > Mine is 4 gb RAM win 7 64 bit. Taking long time. Is it enough
> 
> Mine is 8 gb ram 64-bit os and gpu is very less. It took long time to train like for 1 epoch it took 5 hrs. Imagine 51 to 102 epochs.
> So i asked my frnd to train tge model on her lap. Even her lap is having 8 gb ram but she have 4 gb of gpu. So it took 1 day to train.
> 
> Better you use somebody else's lap with high gpu or try using cloud platforms 

Ok thanks i have G5 server. Will try on that

**Comment by allipilli-harshitha** [comment]
@rpentela  ok do it. Hope for the best

**Comment by rpentela** [comment]
> @rpentela  ok do it. Hope for the best 

Hi I have tried but still val_loss is nan. Did you change batch and subdivisions under testing and training

[net]
#Testing
batch=1
Subdivisions=1
#Training
#batch=64
#subdivisions=16

Did u change the above lines.

**Comment by allipilli-harshitha** [comment]
No i have not changed them.
I have changed only classes in cfg file at 3 places.

**Comment by rpentela** [comment]
> No i have not changed them.
> I have changed only classes in cfg file at 3 places.

Hi I started fresh. I didn't find yolo.h5 in this repository. where did u download it

**Comment by allipilli-harshitha** [comment]
> > No i have not changed them.
> > I have changed only classes in cfg file at 3 places.
> 
> Hi I started fresh. I didn't find yolo.h5 in this repository. where did u download it

1.annotate images with vott
2. Execute annotate.py which is in 1_image_annotation folder 
3. In 2_training folder, download_and_convert.py file you need to run, then you will get yolov3.weights and yolo.h5 files
4. Run train_yolo.py

**Comment by rpentela** [comment]
> > > No i have not changed them.
> > > I have changed only classes in cfg file at 3 places.
> > 
> > Hi I started fresh. I didn't find yolo.h5 in this repository. where did u download it
> 
> 1.annotate images with vott
> 2. Execute annotate.py which is in 1_image_annotation folder 
> 3. In 2_training folder, download_and_convert.py file you need to run, then you will get yolov3.weights and yolo.h5 files
> 4. Run train_yolo.py 

Hi finally I am able to detect.

1) I have trained with width,height as 224×224 in yolov3.cfg classes =2 and input shape as (224,224) in train_yolo.py

Since for this size my pc is not hanging. Otherwise with 416×416 it is hanging

In detector.py also i have mentioned size as 224×224

2) How much confidence score to be considered ? Since for a 3rd class which is not trained and totally different also it is detecting some 58% score.

3) I have clubbed two images of trained classes using mspaint which comes to size 416×416. On this it is detecting only one part of image by changing detector.py. Any idea about this?

**Comment by allipilli-harshitha** [comment]
Hi @rpentela  sry i have no idea about those things because i have not changed any parameters in detector.py. so i have no idea about it.
